### PR TITLE
KMS-667: Update Keywords in ISO-SMAP native records

### PR DESCRIPTION
**Pull request by htranho**
# Overview

### What is the feature?

Update Keywords in ISO-SMAP native records.

### What is the Solution?

Update keywords in following schemes as they appear in ISO-SMAP records, according to correction requests:

sciencekeywords
locations
platforms
instruments
projects
providers
isotopiccategory
productlevelid

### What areas of the application does this impact?

Consumer - Update Keywords in ISO-SMAP native records

# Testing

Run:

```bash
npx vite-node --config vite.config.js scripts/local/run_metadata_correction_iso-smap_smoke.mjs
```

The request fixture currently includes the supported UMM-C correction matrix:

- `sciencekeywords`
- `locations`
- `platforms`
- `instruments`
- `projects`
- `providers`
- `isotopiccategory`
- `productlevelid`

The script prints a JSON summary like:

```json
Applied corrections locally
{
  "fixturePath": "/Users/htranho/repos/kms/scripts/local/fixtures/metadata_correction_service.iso-smap.corrections.json",
  "collectionConceptId": "C1234567223-LOCAL",
  "nativeFormat": "ISO_SMAP",
  "requestedCorrections": 10,
  "appliedCorrections": 10,
  "outputPath": "/Users/htranho/repos/kms/tmp/metadata_correction_service_iso-smap_smoke_output.xml",
  "writeResult": {
    "targetComponent": "cmr-writeback",
    "collectionConceptId": "C1234567223-LOCAL",
    "providerId": null,
    "nativeId": null,
    "nativeFormat": "ISO_SMAP",
    "correctionCount": 10,
    "correctionsAppliedCount": 10,
    "correctedMetadataBytes": 14103,
    "source": "local-smoke",
    "ingestResult": {
      "enabled": false,
      "ingested": false,
      "updated": false
    }
  }
}

```

The important signals are:

- `nativeFormat` is `ISO_SMAP`
- `requestedCorrections` matches the fixture
- `appliedCorrections` matches the expected applied count
- `writeResult.targetComponent` is `cmr-writeback`

The corrected XML is written to:

```bash
tmp/metadata_correction_service_iso-smap_smoke_output.xml
```

See scripts/local/fixtures/metadata_correction_service.iso-smap.corrections.json for requested corrections and compare them with result in tmp/metadata_correction_service_iso-smap_smoke_output.xml

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Summary by CodeRabbit

* **New Features**
  * Implemented real ISO-SMAP metadata corrections (replacing the prior stub) with accurate counts and corrected XML output.
  * Added SMAP-aware ISO format detection and SMAP-safe XPath handling to correctly target edits/deletions.
  * Enhanced ISO editor configurability for keyword-block targeting and added distribution data format corrections.

* **Tests**
  * Expanded Vitest coverage for ISO format detection, SMAP XPath/selection behavior, and ISO-SMAP correction schemes, including edge cases.

* **Chores**
  * Added new local ISO-SMAP fixtures/mocks and a metadata correction smoke script.

### CODE REVIEW

File: serverless/src/shared/Iso19115DomEditor.js [hunk 1]
```diff
@@ -305,6 +331,16 @@ export const ISO_19115_SCHEME_EDITORS = {
   projects: createKeywordBlock('project', {
```
Comment (cgokey) [inline serverless/src/shared/Iso19115DomEditor.js:311]: This is treating gmd:aggregationInfo/gmd:MD_AggregateInformation[@initiativeType='mission'] as a writable project target and rewriting aggregateDataSetIdentifier/.../gmd:code. The mapping workbook says SMAP projects should be handled through gmi:MI_Operation (gov.nasa.esdis.umm.projectshortname), so this path should not be updated and applyIsoSmapMetadataCorrections.Looks like the test is doing this as well, and shouldn't.

File: serverless/src/shared/Iso19115DomEditor.js [hunk 2]
```diff
@@ -158,13 +172,21 @@ export const ISO_19115_SCHEME_EDITORS = {
...
+        gmd:type/gmd:MD_KeywordTypeCode/@codeListValue = 'place' or 
+        gmd:type/gmd:MD_KeywordTypeCode/@codeListValue = 'theme'
+      ]`.replace(/\s+/g, ' ')
     }
   ),
 
   platforms: createKeywordBlock('platform', {
     fieldKeys: ['ShortName', 'LongName'],
     matchKeys: ['ShortName'],
+    nodeXPath: `//gmd:descriptiveKeywords/gmd:MD_Keywords[
```
Comment (cgokey) [inline serverless/src/shared/Iso19115DomEditor.js]: It looks like this broadens locations, platforms, and instruments to match any MD_KeywordTypeCode='theme' block. Since the editor does not also constrain on thesaurus/title, that seems risky because it could update or delete unrelated theme keywords.   Maybe you can match on thesaurusName/CI_Citation/title is the expected vocabulary as well?

File: serverless/src/shared/applyIsoSmapMetadataCorrections.js [hunk 3]
```diff
@@ -1,25 +1,54 @@
...
+      correctedMetadata: undefined,
+      correctionsApplied: [],
+      stubbed: false
+    }
+  }
+
+  // Create editor with SMAP format - automatically transforms XPath expressions
+  const editor = createIso19115Editor(metadataPayload, { format: 'SMAP' })
+
+  const applied = corrections.reduce((acc, correction) => {
```
Comment (cgokey) [inline serverless/src/shared/applyIsoSmapMetadataCorrections.js:29]: This looks like it inherits the generic dataCentre keyword-block editor from Iso19115DomEditor.js. Since the workbook’s SMAP data-center mapping is role-based and would require XML-context-aware handling, I think providers should be explicitly unsupported for SMAP like we do with the others that are role based.

File: serverless/src/shared/Iso19115DomEditor.js (hunk 2 above)
Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js]: Agree, going to remove the condition 'theme' for locations, platforms, and instruments (can't use title because in existing SMAP records have the same title 'NASA/GCMD Earth Science Keywords' for all of those schemes)

File: serverless/src/shared/applyIsoSmapMetadataCorrections.js (hunk 3 above)
Comment (htranho) [reply serverless/src/shared/applyIsoSmapMetadataCorrections.js:29]: SMAP doesn't have 'dataCentre'. This query condition should stay for the CWIC and NOOA 19115 records.

Comment (htranho) [reply serverless/src/shared/applyIsoSmapMetadataCorrections.js:29]: about issue with data center in SMAP: The role base paths is the mapping file are not those in '//gmd:descriptiveKeywords/gmd:MD_Keywords', so we are fine here for both MENDS and SMAP. The code is only looking for those in //gmd:descriptiveKeywords/gmd:MD_Keywords to fix and not in organization.

File: serverless/src/shared/Iso19115DomEditor.js (hunk 1 above)
Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js:311]: Will remove this path (//gmd:aggregationInfo/gmd:MD_AggregateInformation[@initiativeType='mission'] ) from 'projects'

Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js:311]: removed

File: serverless/src/shared/Iso19115DomEditor.js (hunk 2 above)
Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js]: Now removed

File: serverless/src/shared/Iso19115DomEditor.js [hunk 4]
```diff
@@ -325,6 +331,15 @@ export const ISO_19115_SCHEME_EDITORS = {
       {
         path: '//gmi:acquisitionInformation/gmi:MI_AcquisitionInformation/gmi:operation/gmi:MI_Operation/gmi:identifier/gmd:MD_Identifier[gmd:codeSpace/gco:CharacterString="gov.nasa.esdis.umm.projectshortname"]/gmd:description/gco:CharacterString',
         getValue: ({ correction }) => correction.newLongName || ''
+      },
+      {
+        path: 'gmd:aggregateDataSetIdentifier/gmd:MD_Identifier/gmd:code/gco:CharacterString',
```
Comment (cgokey) [inline serverless/src/shared/Iso19115DomEditor.js]: This stil includes gmd:aggregateDataSetIdentifier/.../gmd:code, but the workbook’s writable SMAP project target is gmi:MI_Operation, not the aggregate mission block. Since this path is also "relative", it does not appear to be an active SMAP write targe (right?), so I think we should remove it rather than keep an unsupported mapping around.

Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js]: Good catch, now removed

**Review by cgokey — APPROVED** [review]

**Comment by william-valencia** [comment]
Should there be a DataFormat field?
[=> /gmd:distributionFormat xlink:href="FileDistributionInformation_<Block Number>"/gmd:MD_Format/gmd:name/gco:CharacterString

**Review by eudoroolivares2016 — APPROVED** [review]

**Comment by htranho** [comment]
> Should there be a DataFormat field? [=> /gmd:distributionFormat xlink:href="FileDistributionInformation_"/gmd:MD_Format/gmd:name/gco:CharacterString

Now added dataformat to both ISO-MENDS and ISO-SMAP.

File: serverless/src/shared/Iso19115DomEditor.js [hunk 5]
```diff
@@ -338,10 +344,23 @@ export const ISO_19115_SCHEME_EDITORS = {
...
+    replace: [
+      {
+        fieldPath: '.',
+        source: {
+          type: 'computed',
+          getValue: ({ correction }) => correction.newKeywordObject.Value
+        }
+      }
+    ]
   }),
```
[the remark this answers is not in the document]
Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js:369]: Fixed

File: serverless/src/shared/Iso19115DomEditor.js [hunk 6]
```diff
@@ -338,10 +344,27 @@ export const ISO_19115_SCHEME_EDITORS = {
 
       return LongName ? `${ShortName} > ${LongName}` : ShortName
     }
-    // Additional paths
-    // additionalPaths: [
-    //   '//gmd:CI_ResponsibleParty/gmd:organisationName/gco:CharacterString'
-    // ]
+  }),
+
+  dataformat: leafScheme({
```
Comment (cgokey) [inline serverless/src/shared/Iso19115DomEditor.js:349]: I think this should be ShortName, no? ```suggestion
dataformat: leafScheme({
  // Target the wrapper block so deletion removes the entire format entry cleanly
  nodeXPath: '//gmd:distributionInfo/gmd:MD_Distribution/gmd:distributionFormat',
  find: {
    getNodeValueObject: ({ node, editor }) => {
      const charStringNode = editor.selectNodes('.//gmd:name/gco:CharacterString', node)[0]

      return { ShortName: charStringNode?.textContent?.trim() || '' }
    }
  },
  replace: [
    {
      fieldPath: 'gmd:MD_Format/gmd:name/gco:CharacterString',
      source: {
        type: 'computed',
        getValue: ({ correction }) => correction.newKeywordObject.ShortName
      }
    }
  ]
}),```

Comment (htranho) [reply serverless/src/shared/Iso19115DomEditor.js:349]: Now fixed
